dex/farm/src/lib.rs

```diff
@@ -613,4 +613,207 @@ pub trait Farm:
+
+        let old_position = payments.get(0).unwrap();
+        assert!(self, old_position.amount != 0u64, b"bad farm amount");
+
+        assert!(
+            self,
+            old_position.token_identifier == config.old_farm_token_id,
+            b"bad farm token id"
+        );
+
+        let reward = payments.get(1).unwrap();
+        assert!(self, reward.amount != 0u64, b"bad reward amount");
+
+        let reward_token_id = self.reward_token_id().get();
+        assert!(
+            self,
+            reward.token_identifier == reward_token_id,
+            b"bad reward token id"
+        );
+
+        // The actual work starts here
+        self.reward_reserve().update(|x| *x += &reward.amount);
+
+        let old_attrs: FarmTokenAttributes<Self::Api> = self
+            .blockchain()
+            .get_esdt_token_data(
+                &self.blockchain().get_sc_address(),
+                &old_position.token_identifier,
+                old_position.token_nonce,
+            )
+            .decode_attributes()
+            .unwrap();
+
+        // Do not call burn_farm_tokens since this farm tokens belong to other contract
+        // which already updated its farm token supply counter.
+        self.send().esdt_local_burn(
+            &old_position.token_identifier,
+            old_position.token_nonce,
+            &old_position.amount,
+        );
+
+        let new_pos_token_id = self.farm_token_id().get();
+        let new_pos_amount = old_position.amount;
+
+        // Use this function because it also updates the farm token supply for this contract instance.
+        let new_pos_nonce = self.mint_farm_tokens(&new_pos_token_id, &new_pos_amount, &old_attrs);
+
+        let orig_caller = orig_caller_opt
+            .into_option()
+            .unwrap_or_else(|| caller.clone());
+
+        // Use this function since it works regardless of wasm ocasional unalignment.
+        self.transfer_execute_custom(
+            &orig_caller,
+            &new_pos_token_id,
+            new_pos_nonce,
+            &new_pos_amount,
+            &OptionalArg::None,
+        );
+
+        EsdtTokenPayment::new(new_pos_token_id, new_pos_nonce, new_pos_amount)
+    }
+
+    // Each farm that will be migrated and the newer version to which we migrate to
+    // will have to be configured using this function.
+    #[only_owner]
+    #[endpoint(setFarmConfiguration)]
+    fn set_farm_configuration(
+        &self,
+        is_old: bool,
+        old_farm_address: ManagedAddress,
+        old_farm_token_id: TokenIdentifier,
+        new_farm_address: ManagedAddress,
+    ) {
+        self.farm_configuration().set(&FarmContractConfig {
+            is_old,
+            old_farm_address,
+            old_farm_token_id,
+            new_farm_address,
+        });
+    }
+
+    // We also need to get the rps and transfer it to the new SC.
+    #[only_owner]
+    #[endpoint(stopRewardsAndGetRps)]
+    fn stop_rewards_and_get_rps(&self) -> BigUint {
```

Comment:
I think it would be better to directly have a dunction of migrate rewards and RPS. So there would be no need to 2 transactions from the owner. Only one. And contract calling another contract setting up the values.

Reply:
Done. I have written it.

Reply:
We cannot go with this code, as we will move from contract v1.2 to v1.4 split farms where user decides in which farm he wants to go - the rewards per share cannot be used. We need to make enterFarmWithOriginalEpoch.